pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,201 @@
+	}
+
+	// Support retrieving logs for older deployments
+	switch {
+	case deployLogOpts.Version == nil:
+		// Latest
+	case *deployLogOpts.Version <= 0 || *deployLogOpts.Version > config.LatestVersion:
+		// Invalid version
+		return nil, errors.NewBadRequest(fmt.Sprintf("invalid version for deploymentConfig %q: %d", config.Name, *deployLogOpts.Version))
+	default:
+		desiredVersion = *deployLogOpts.Version
+	}
+
+	// Get desired deployment
+	targetName := deployutil.DeploymentNameForConfigVersion(config.Name, desiredVersion)
+	target, err := r.DeploymentGetter.ReplicationControllers(namespace).Get(targetName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for deployment status; if it is new or pending, we will wait for it. If it is complete,
+	// the deployment completed successfully and the deployer pod will be deleted so we will return a
+	// success message. If it is running or failed, retrieve the log from the deployer pod.
+	status := deployutil.DeploymentStatusFor(target)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s is in %s state. No logs to retrieve yet.", deployutil.LabelForDeployment(target), status)
+			return &genericrest.LocationStreamer{}, nil
+		}
+		glog.V(4).Infof("Deployment %s is in %s state, waiting for it to start...", deployutil.LabelForDeployment(target), status)
+
+		if err := waitForDeployment(r.DeploymentGetter, target, r.Timeout); err != nil {
+			return nil, err
+		}
+	case deployapi.DeploymentStatusComplete:
+		// Deployer pod has been deleted, no logs to retrieve
+		glog.V(4).Infof("Deployment %s was successful so the deployer pod is deleted. No logs to retrieve.", deployutil.LabelForDeployment(target))
+		return &genericrest.LocationStreamer{}, nil
+	}
+
+	// Setup url of the deployer pod
+	deployPodName := deployutil.DeployerPodNameForDeployment(target.Name)
+	logOpts := &kapi.PodLogOptions{
+		Follow: deployLogOpts.Follow,
+	}
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       deployLogOpts.Follow,
+	}, nil
+}
+
+// waitForDeployment waits until the passed deployment is Running, Complete, or Failed
+func waitForDeployment(c kclient.ReplicationControllersNamespacer, deployment *kapi.ReplicationController, timeout time.Duration) error {
+	fieldSelector := fields.Set{"metadata.name": deployment.Name}.AsSelector()
+	w, err := c.ReplicationControllers(deployment.Namespace).Watch(labels.Everything(), fieldSelector, deployment.ResourceVersion)
+	if err != nil {
+		return err
+	}
+	defer w.Stop()
+
+	done := make(chan struct{})
+	errchan := make(chan error)
+	go func(ch <-chan watch.Event) {
```

Comment:
See my factoring of this in the Binary Build pull request - this can be made much, much simpler.